addons/crm_addons/controllers/main.py

```diff
@@ -0,0 +1,566 @@
+            values['login'] = request.session.get('auth_login')
+
+        if not odoo.tools.config['list_db']:
+            values['disable_database_manager'] = True
+
+        response = request.render('crm_addons.login', values)
+        response.headers['X-Frame-Options'] = 'DENY'
+        return response
+
+    def _generate_code(self, uid, time_valid):
+        secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        encoded_payload = ('%s.%s' % (uid, time_valid)).encode()
+        signature = hmac.new(secret.encode(), encoded_payload, sha256).hexdigest()
+
+        return '%s.%s.%s' % (uid, time_valid, signature)
+
+    def _check_signature(self, code):
+        secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        uid, time_valid, signature = code.split('.')
+        check_signature = hmac.new(secret.encode(), ('%s.%s' % (uid, time_valid)).encode(), sha256).hexdigest()
+        if not hmac.compare_digest(check_signature, signature):
+            return False
+
+        if int(time_valid) < int(datetime.utcnow().timestamp()):
+            return False
+
+        return uid
+
+    @http.route('/mail_client_extension/token', type='http', csrf=False, auth="none", cors="*")
+    def exchange_code_with_token(self, **kw):
+        code = request.params['code']
+
+        uid = self._check_signature(code)
+        if not uid:
+            return json.dumps({
+                "error": "Invalid code",
+            })
+
+        # GENERATE TOKEN
+        header = json.dumps({
+            "alg": "SHA256",
+            "typ": "JWT"
+        })
+        # TODO see JWT claims: https://exante.eu/clientsarea/tutorials/jwt/#payload
+        payload = json.dumps({
+            "user_id": uid,
+            'time_valid': int((datetime.utcnow() + timedelta(days=30)).timestamp())
+        })
+
+        b64_encoded_string = "%s.%s" % (base64.b64encode(header.encode()).decode(), base64.b64encode(payload.encode()).decode())
+        secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        signature = hmac.new(secret.encode(), b64_encoded_string.encode(), sha256).hexdigest()
+        b64_encoded_signature = base64.b64encode(signature.encode()).decode()
+        return json.dumps({
+            "access_token": b64_encoded_string + "." + b64_encoded_signature,
+            "expires_in": timedelta(days=30).total_seconds(), # For gmail.
+        })
+
+    def _iap_enrich(self, domain):
+        #normalized_email = tools.email_normalize(partner.email)
+        #if not normalized_email:
+        #    return {}
+        #emails = {0: normalized_email.split('@')[1]}
+        domains = {0: domain}
+
+        enriched_data = {}
+        #enriched_data["enrichment_error"] = {'type': 'none', 'info': ''}
+        try:
+            response = request.env['iap.enrich.api'].with_user(request.env.user)._request_enrich(domains)
```

Comment:
The env's user is already the env's user, that doesn't seem useful? As far as I know, all `with_user` does is set the parameter as the env's user (and re-sets the `su` flag to `False` but then if that's the goal `.sudo(False)` is clearer).